lib/sequent/core/event.rb

```diff
@@ -24,14 +22,22 @@ def initialize(args = {})
 
       def payload
         result = {}
-        instance_variables.reject { |k| payload_variables.include?(k.to_s) }.each do |k|
+        instance_variables
+          .reject { |k| payload_variables.include?(k.to_s) }
+          .select { |k| attributes.keys.include?(to_attribute_name(k))}
+          .each do |k|
           result[k.to_s[1 .. -1].to_sym] = instance_variable_get(k)
         end
         result
       end
       protected
       def payload_variables
-        %w{@aggregate_id @sequence_number @created_at @underscored}
+        %w{@aggregate_id @sequence_number @created_at}
+      end
+
+      private
+      def to_attribute_name(instance_variable_name)
+        instance_variable_name.to_s.gsub('@', '').to_sym
```

Comment:
Only remove the leading `@` instead of replacing all with the empty string. The slice operator (`s[1..-1]`) also works on symbols, so the the `to_s` can be removed (you still need to `to_sym` though).